### PR TITLE
Wip - Puts issue #4118 under test - test min similarities

### PR DESCRIPTION
**Pull request by doublethefish**
## Steps

- [ ] Add yourself to CONTRIBUTORS if you are a new contributor.
- [ ] Add a ChangeLog entry describing what your PR does.
- [ ] If it's a new feature or an important bug fix, add a What's New entry in `doc/whatsnew/<current release.rst>`.
- [ ] Write a good description on what the PR does.

## Description

Puts issue #4118 under test …
1b477cd
Taken in isolation, this commit may appear heavy-weight. The intent here
is to have a linter run and retain all state information about that run
so we can inspect what happened.

What this commit shows is that the checkers the linter knows about have
the right config, but still multi-job [MJ] runs are incorrect.
    see: assert checker.config.min_similarity_lines == min_similarity_lines

I think the MJ failiures are because SimilarChecker.reduce_map_data()
constructs a new SimilarChecker and then the linter doesn't call
set_option().

The process by which the set_option() calls are made in the normal run
is opaque to me, but perhaps someone else has a better idea(?).

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :bug: Bug fix  |
| ✓  | :sparkles: New feature |
| ✓  | :hammer: Refactoring  |
| ✓  | :scroll: Docs |

### CODE REVIEW

File: tests/checkers/unittest_similar.py [hunk 1]
```diff
@@ -248,9 +252,11 @@ def test_get_map_data():
     linter.register_checker(similar.SimilarChecker(linter))
 
     source_streams = (
-        str(INPUT / "similar_lines_a.py"),
-        str(INPUT / "similar_lines_b.py"),
+        SIMILAR_A,
+        SIMILAR_B,
     )
+    for fname in source_streams:
+        assert os.path.exists(fname), f"File not found! {fname}"
```
Comment (Pierre-Sassoulas) [inline tests/checkers/unittest_similar.py:259]:
```suggestion
        assert fname.exists(), f"File not found! {fname}"
```

**Review by Pierre-Sassoulas — COMMENTED** [review]
Thank you for working on this one, you're really starting to feel like the multi-processing expert around here 😄

**Comment by shvenkat** [comment]
Just saw this... I came across the same issue, and proposed a fix in #4178.

**Comment by doublethefish** [comment]
> Just saw this... I came across the same issue, and proposed a fix in #4178.

Thank you so much for highlighting that, I had completely missed the PR. Legend. :)

**Comment by doublethefish** [comment]
I am happy for these tests to go in now. Thoughts?

File: tests/checkers/unittest_similar.py [hunk 2]
```diff
@@ -375,3 +380,148 @@ def test_get_map_data():
...
+            assert checker.config.min_similarity_lines == min_similarity_lines
+            assert not hitonce
+            hitonce = True
+    assert hitonce
+
+    return exit_code, stdout
+
+
+@pytest.mark.xfail
+@pytest.mark.parametrize("min_similarity_lines", [0, 1, 4, 1000])
```
Comment (Pierre-Sassoulas) [inline tests/checkers/unittest_similar.py:451]: 👍

File: tests/checkers/unittest_similar.py [hunk 3]
```diff
@@ -375,3 +380,148 @@ def test_get_map_data():
...
+    assert exit_single == exit_multi
+
+
+def _simcheck_with_cli(num_jobs, min_similarity_lines, tmp_path):
+    """Tests check_parallel passes the cli options to sub-workers
+
+    Partially tests bug #4118 and #4173"""
+    config_file = tmp_path / "setup.cfg"
+
+    source_streams = [
```
Comment (Pierre-Sassoulas) [inline tests/checkers/unittest_similar.py:481]: Could we create a source_streams fixture ? I've seen this part + the check three times :)

File: tests/test_config.py [hunk 4]
```diff
@@ -1,23 +1,53 @@
...
+
+from test_self import _patch_streams
 
 import pylint.lint
 
 
-def check_configuration_file_reader(config_file):
-    """Initialize pylint with the given configuration file and check that
-    what we initialized the linter with what was expected.
+def run_with_config_file(config_file, args):
```
Comment (Pierre-Sassoulas) [inline tests/test_config.py:10]:
```suggestion
def run_with_config_file(config_file;Union[Path, str], args: List[str]) -> Tuple[runner?, int, str]:
```
We do not have to maintain python 3.5 we can put typing in signature directly :)

**Review by Pierre-Sassoulas — APPROVED** [review]
Nice tests, they will definitely help with multi-processing issues.

File: tests/checkers/unittest_similar.py (hunk 3 above)
Comment (doublethefish) [reply tests/checkers/unittest_similar.py:481]: Sure, if I get time :)

**Comment by bentheadset** [comment]
The open bugs only appear to reference py 2.x - I see the same issue on py3.8.  Is this expected to fix both, or should i open a new bug?

also, I see this is `! wip` and approved - i'm not familiar with `pylint` release cadence, approximately when would this end up in pypi?

**Comment by Pierre-Sassoulas** [comment]
@bentheadset 2.7.0 in the issue is a version of pylint, not python we're only supporting python 3.6 to 3.9 right now. Regarding the MR we'll merge it once the test pass. And it's going to be in the next release of pylint, we don't have a set schedule for that.

**Comment by Pierre-Sassoulas** [comment]
@doublethefish should we close this MR in favor of #4175 ?

**Comment by Pierre-Sassoulas** [comment]
I'm closing in order to cleanup the PR lists please do not hesitate to reopen a new one if necessary.

**Pull request closed without merging**